pkg/kds/mux/session.go

```diff
@@ -14,65 +17,115 @@ type Session interface {
 	PeerID() string
 	Done() <-chan struct{}
 	Error() error
+	Close()
 }
 
 type session struct {
 	peerID       string
+	closing      chan struct{}
 	done         chan struct{}
 	err          chan error
 	serverStream *kdsServerStream
 	clientStream *kdsClientStream
-	closed       int32
+	wg           sync.WaitGroup
+	cancel       func()
 }
 
-func NewSession(peerID string, stream MultiplexStream, stop <-chan struct{}) Session {
+type MultiplexStream interface {
+	Send(*mesh_proto.Message) error
+	Recv() (*mesh_proto.Message, error)
+	Context() context.Context
+}
+
+func NewSession(peerID string, stream MultiplexStream) Session {
+	ctx, cancel := context.WithCancel(stream.Context())
 	s := &session{
 		peerID: peerID,
 		done:   make(chan struct{}, 1),
 		err:    make(chan error, 1),
 		serverStream: &kdsServerStream{
-			requests:        make(chan *envoy_sd.DiscoveryRequest, 1),
-			MultiplexStream: stream,
+			bufferStream: newBufferStream(ctx),
 		},
 		clientStream: &kdsClientStream{
-			responses:       make(chan *envoy_sd.DiscoveryResponse, 1),
-			MultiplexStream: stream,
+			bufferStream: newBufferStream(ctx),
 		},
-		closed: int32(0),
+		closing: make(chan struct{}),
+		wg:      sync.WaitGroup{},
+		cancel:  cancel,
 	}
 	go func() {
-		defer s.close()
-		if err := s.handle(stream, stop); err != nil {
-			s.err <- err
-		}
+		s.handleRecv(stream)
+	}()
+	s.wg.Add(1)
```

Comment:
Probably because this depends on the receive side getting closed by the GRPC runtime. But I think it would be a lot clearer if the wait was symmetric.